apps/dotcom-worker/src/lib/TLDrawDurableObject.ts

```diff
@@ -376,48 +399,29 @@ export class TLDrawDurableObject extends TLServer {
 		}
 	}
 
-	_isPersisting = false
 	_lastPersistedClock: number | null = null
+	_persistQueue = createPersistQueue(async () => {
+		// check whether the worker was woken up to persist after having gone to sleep
+		if (!this._room) return
+		const slug = this.documentInfo.slug
+		const room = await this.getRoom()
+		const clock = room.getCurrentDocumentClock()
+		if (this._lastPersistedClock === clock) return
+		if (this._isRestoring) return
+
+		const snapshot = JSON.stringify(room.getCurrentSnapshot())
+
+		const key = getR2KeyForRoom(slug)
+		await Promise.all([
+			this.r2.rooms.put(key, snapshot),
+			this.r2.versionCache.put(key + `/` + new Date().toISOString(), snapshot),
+		])
+		this._lastPersistedClock = clock
+	}, PERSIST_INTERVAL_MS / 2)
```

Comment:
ooc, why does this get divided by two?

Reply:
the PERSIST_INTERVAL_MS const is kinda for an 'outer' loop, implemented with the alarm scheduler. This bit is kinda for an 'inner' loop that allows `await this.persistToDatabase()` to wait on the whole persist queue to drain rather than just one persist operation. (we do this when the room shuts down). I was worried that if the two loops use the same interval it might be possible for the outer loop to keep the queue backed up for as long as document changes are coming in.

now I'm writing this out it seems unlikely but 🤷🏼 can't hurt

Reply:
added a comment explaining